File: 383-ransom-note/383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
const canConstruct = function(ransomNote, magazine) {
    
    const ransomMap = {};
    const magMap = {};
    
    for (let i = 0, n = ransomNote.length; i < n; i ++) {
        const char = ransomNote[i];
        
        if (ransomMap[char]) {
            ransomMap[char] ++;
        } else {
            ransomMap[char] = 1; 
        }
    }
    
    for (let i = 0, n = magazine.length; i < n; i ++) {
        const char = magazine[i];
        
        if (magMap[char]) {
            magMap[char] ++;
        } else {
            magMap[char] = 1; 
        }
        
    }
    
    for (const char in ransomMap) {
        if (!magMap[char] || ransomMap[char] > magMap[char]) return false;
    }
    
    return true;
};

File: 383-ransom-note/383-ransom-note.rb
# @param {String} ransom_note
# @param {String} magazine
# @return {Boolean}
def can_construct(ransom_note, magazine)
   mag_chars_count = count_char(magazine)
    
   ransom_note.each_char do |char|
       mag_chars_count[char] -= 1
       return false if mag_chars_count[char] < 0
   end
    
   true
end


def count_char(str)
    count = Hash.new(0)
    str.each_char do |char|
        count[char] += 1
    end
    count
end